pytorch_lightning/trainer/progress.py

```diff
@@ -0,0 +1,180 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from dataclasses import dataclass, field
+from typing import Optional
+
+
+@dataclass
+class Progress:
+    """
+    Basic dataclass to track event progress.
+
+    Args:
+        ready: Intended to track the number of events ready to start.
+        started: Intended to be incremented after ``on_*_start`` completes.
+        processed: Intended to be incremented after the event is processed.
+        completed: Intended to be incremented after ``on_*_end`` completes.
+
+    Attributes set to ``None`` are treated as unused and are restricted.
+    """
+    ready: Optional[int] = 0
+    started: Optional[int] = 0
+    processed: Optional[int] = 0
+    completed: Optional[int] = 0
+
+    def reset(self) -> None:
+        if self.ready is not None:
+            self.ready = 0
+        if self.started is not None:
+            self.started = 0
+        if self.processed is not None:
+            self.processed = 0
+        if self.completed is not None:
+            self.completed = 0
+
+    def __setattr__(self, key: str, value: int) -> None:
+        if getattr(self, key, 0) is None:
+            raise AttributeError(f"The '{key}' attribute is meant to be unused")
+        return super().__setattr__(key, value)
+
```

Comment:
this means at construction we need to null out the fields we want to mark as unused. for users, indicating this behavior change with the API based on types feels hidden. this also means if anyone extends this class and really wants an optional field which is initially None but later tracked for their loop, they also need to override `__setattr__`

since these are brand new structs, should we instead offer per-use case structs that contain only the fields we need? 

for example: 
- a BatchProgress that has ready/started/processed/completed + a BatchLoopProgress which has total + current BatchProgress members
- an OptimizationProgress which has processed steps + OptimizationLoopProgress which has total + current OptimizationProgress members?
- an EpochProgress which has started + completed epoch counts

Reply:
> indicating this API behavior change based purely on types feels is not intuitive.

I agree it's not the clearest but to me seems like an okay tradeoff to avoid the code duplication.

> this also means if anyone extends this class

I don't think there is a reason to extend the `Progress` class in particular. `BaseProgress` and so on might but they don't contain this "unused" logic.

> since these are brand new structs, should we instead offer per-use case structs that contain only the fields we need?

> would it be easier to introduce a separate optimization progress struct which doesn't carry all these fields, and if there is demand for them later, then we can unify using the same struct?

I'm open to it, but do you have an idea how? Because this wouldn't work well:

```python
@dataclass
class Progress:  # name?
    ready: int = 0
    started: int = 0
    processed: int = 0
    completed: int = 0


@dataclass
class Progress1:  # name?
    ready: int = 0
    started: int = 0
    completed: int = 0


@dataclass
class Progress2:  # name?
    ready: int = 0
    completed: int = 0

@dataclass
class OptimizationProgress:
    optimizer: BaseProgress = BaseProgress(total=Progress1(), current=Progress1())
    scheduler: BaseProgress = BaseProgress(total=Progress2(), current=Progress2())
    zero_grad: BaseProgress = BaseProgress(total=Progress1(), current=Progress1())
```

There are several issues with this structure: 

- Duplicated docstrings between classes
- Duplicated helper functions between classes
- Typing complaints as there is no shared class
- BaseProgress assumes the 4 attributes are available

> do we think optimization progress would use these fields later?
Maybe. I couldn't think of a use for them but they could be useful in the future.